ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/DbChangelog.hs

```diff
@@ -0,0 +1,427 @@
+}
+
+nextState :: (DbChangelog TestLedger) -> TestLedger DiffMK
+nextState dblog = TestLedger
+            { tlTip = pointAtSlot $ nextSlot (getTipSlot old)
+            , tlUtxos = ApplyDiffMK $ emptyUtxoDiff
+            }
+  where
+    old = unDbChangelogState $ either id id $ AS.head (changelogVolatileStates dblog)
+    nextSlot Origin = At 1
+    nextSlot (At x) = At (x + 1)
+
+nExtensions :: Int -> DbChangelog TestLedger -> DbChangelog TestLedger
+nExtensions n dblog = (iterate extend dblog) !! n
+  where extend dblog' = extendDbChangelog dblog' (nextState dblog')
+
+deriving instance Show (TestLedger EmptyMK)
+deriving instance Show (TestLedger DiffMK)
+
+instance GetTip (TestLedger EmptyMK) where
+  getTip = tlTip
+
+instance GetTip (TestLedger DiffMK) where
+  getTip = castPoint . tlTip
+
+data H = H deriving (Eq, Ord, Show, Generic)
+deriving anyclass instance NoThunks H
+type instance HeaderHash (TestLedger mk) = H
+
+instance StandardHash (TestLedger EmptyMK)
+
+deriving instance Eq (TestLedger EmptyMK)
+deriving instance Eq (LedgerTables TestLedger DiffMK)
+deriving instance Eq (LedgerTables TestLedger ValuesMK)
+
+instance ShowLedgerState (LedgerTables TestLedger) where
+  showsLedgerState _ (TestTables t) = showString "TestTables " . shows t
+
+-- REVIEW: What to do with this orphan instance?
+instance Show (ApplyMapKind' mk' Key Int) where
+  show ap = showsApplyMapKind ap ""
+
+instance ShowLedgerState TestLedger where
+  showsLedgerState _ (TestLedger {tlUtxos, tlTip}) =
+    showString "TestLedger" . showSpace . showString "{" . shows tlUtxos .
+    showCommaSpace . shows tlTip . showString "}"
+
+instance TableStuff TestLedger where
+  data LedgerTables TestLedger mk = TestTables { unTestTables :: ApplyMapKind mk Key Int }
+  projectLedgerTables = TestTables . tlUtxos
+  withLedgerTables st (TestTables x) = st { tlUtxos = x }
+  pureLedgerTables = TestTables
+  mapLedgerTables f (TestTables x) = TestTables (f x)
+  traverseLedgerTables f (TestTables x) = TestTables <$> f x
+  zipLedgerTables f (TestTables x) (TestTables y) = TestTables (f x y)
+  zipLedgerTables2 f (TestTables x) (TestTables y) (TestTables z) = TestTables (f x y z)
+  zipLedgerTablesA f (TestTables x) (TestTables y) = TestTables <$> f x y
+  zipLedgerTables2A f (TestTables x) (TestTables y) (TestTables z) = TestTables <$> f x y z
+  foldLedgerTables f (TestTables x) = f x
+  foldLedgerTables2 f (TestTables x) (TestTables y) = f x y
+  namesLedgerTables = TestTables $ NameMK "TestTables"
+
+deriving instance Eq (LedgerTables TestLedger SeqDiffMK)
+
+data DbChangelogTestSetup = DbChangelogTestSetup
+  { operations          :: [Operation TestLedger]
+  , originalDbChangelog :: DbChangelog TestLedger
+  }
+
+data Operation l = Extend (l DiffMK) | Prune SecurityParam
+deriving instance Show (l DiffMK) => Show (Operation l)
+
+data DbChangelogTestSetupWithRollbacks = DbChangelogTestSetupWithRollbacks
+  { testSetup :: DbChangelogTestSetup
+  , rollbacks :: Int
```

Comment:
Do we need this structure? Why not making the rollback field a parameter of the property.

Reply:
That might be my lack of knowledge. The rollbacks needs to be at most the length of the changelog (or volatile states), otherwise the rollback function throws.